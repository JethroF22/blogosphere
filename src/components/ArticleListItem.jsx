import React from "react";
import moment from "moment";

const ArticleListItem = props => (
  <div className="uk-card uk-card-default uk-card-hover uk-margin-medium-bottom">
    <div className="uk-card-header">
      <h3 className="article-list__title">{props.post.title}</h3>
      <p className="article-list__subtitle">
        Published on {moment(props.post.createdAt).format("MMMM Do YYYY")}
      </p>
    </div>
    <div className="uk-card-media-bottom">
      <img
        className="article-list__image"
        uk-img={`dataSrc: ${props.post.coverPhotoURL}`}
      />
    </div>
  </div>
);

export default ArticleListItem;
