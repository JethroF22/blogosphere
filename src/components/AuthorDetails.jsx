import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faUserCircle } from "@fortawesome/free-solid-svg-icons";

const AuthorDetails = props => (
  <div className="details">
    {props.photo ? (
      <div className="details__left">
        <img src={props.photo} />
      </div>
    ) : (
      <FontAwesomeIcon icon={faUserCircle} size="4x" />
    )}
    <div className="details__right">
      <p>Published by {props.username}</p>
      <button
        className="uk-button uk-button-default button"
        onClick={props.followed ? null : props.followAuthor}
        disabled={props.followed}
      >
        {props.followed ? `Following ${props.username}` : `Followed`}
      </button>
    </div>
  </div>
);

export default AuthorDetails;
