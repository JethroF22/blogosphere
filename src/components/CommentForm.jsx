import React, { Component } from "react";
import { connect } from "react-redux";

import { createComment } from "../actions/blog";

class CommentForm extends Component {
  state = {
    commentBody: ""
  };

  onCommentChange = e => {
    const commentBody = e.target.value;
    this.setState(() => ({ commentBody }));
  };

  onSubmit = e => {
    e.preventDefault();

    this.props
      .createComment(
        this.state.commentBody,
        this.props.postID,
        this.props.token
      )
      .then(slug => {
        if (this.props.actionStatus === "Action successful") {
          this.setState(() => ({
            commentBody: ""
          }));
        }
      });
  };

  render() {
    return (
      <div className="form">
        <form className="uk-form-stacked" onSubmit={this.onSubmit}>
          <label htmlFor="commentBody" className="form__label">
            Leave a comment
          </label>
          <input
            type="text"
            name="commentBody"
            value={this.state.commentBody}
            onChange={this.onCommentChange}
            placeholder="Let the author know what you think..."
            className={"uk-textarea form__textarea"}
          />
          <br />
          <button
            type="submit"
            className="uk-button uk-button-default button"
            disabled={this.state.commentBody ? false : true}
          >
            Submit comment
          </button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  token: state.auth.token,
  postID: state.blog.currentPost._id,
  actionStatus: state.status.status
});

const mapDispatchToProps = dispatch => ({
  createComment: (commentBody, postID, token) =>
    dispatch(createComment(commentBody, postID, token))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CommentForm);
