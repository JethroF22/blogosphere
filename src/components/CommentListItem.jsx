import React from "react";
import moment from "moment";

const CommentListItem = props => (
  <div>
    <h3>
      {props.comment.username} ({moment(props.comment.timestamp).format(
        "Do MMM"
      )})
    </h3>
    <p>{props.comment.body}</p>
  </div>
);

export default CommentListItem;
