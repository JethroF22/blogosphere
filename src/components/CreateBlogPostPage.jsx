import React, { Component } from "react";
import { connect } from "react-redux";
import Textarea from "react-textarea-autosize";

class CreateBlogPostPage extends Component {
  state = {
    body: "",
    title: "",
    coverPhotoURL: ""
  };

  render() {
    return (
      <div>
        <h1>Create a Blog Post</h1>
        <form>
          <label htmlFor="title">Title: </label>
          <input
            type="text"
            name="title"
            value={this.state.title}
            onChange={this.onTitleChange}
          />
          <br />
          <label htmlFor="body">Body: </label>
          <Textarea
            name="body"
            onChange={this.onBodyChange}
            value={this.state.body}
          />
          <br />
          <label htmlFor="coverPhoto">Cover Photo: </label>
          <input
            type="text"
            name="coverPhoto"
            value={this.state.coverPhotoURL}
            onChange={this.onUsernameChange}
            placeholder="Image URL"
          />
          <br />
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  createArticle: (article, token) => dispatch(createArticle(article, token))
});

export default connect(
  undefined,
  mapDispatchToProps
)(CreateBlogPostPage);
