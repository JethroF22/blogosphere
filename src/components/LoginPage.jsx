import React, { Component, Fragment } from "react";
import { connect } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSpinner } from "@fortawesome/free-solid-svg-icons";
import _ from "lodash";
import validator from "validator";

import { startAuthentication } from "../actions/auth";

export class LoginPage extends Component {
  state = {
    password: "",
    email: "",
    errors: {}
  };

  onEmailChange = e => {
    const email = e.target.value;
    this.setState(prevState => ({
      email,
      errors: {
        ...prevState.errors,
        email: ""
      }
    }));
  };

  onPasswordChange = e => {
    const password = e.target.value;
    this.setState(prevState => ({
      password,
      errors: {
        ...prevState.errors,
        password: ""
      }
    }));
  };

  onSubmit = e => {
    e.preventDefault();

    const isValidEmail =
      this.state.email && validator.isEmail(this.state.email);
    const isValidPassword = this.state.password;

    if (isValidPassword && isValidEmail) {
      this.props
        .login(_.pick(this.state, ["email", "password"]))
        .then(token => {
          if (this.props.actionStatus === "Action successful") {
            localStorage.setItem("token", token);
            this.props.history.push("/");
          }
        });
    } else {
      const errors = {};
      if (!isValidEmail) {
        errors.email =
          "Please enter the email you used to register your account";
      }
      if (!isValidPassword) {
        errors.password = "Please enter your password";
      }
      this.setState(() => ({ errors }));
    }
  };

  render() {
    return (
      <div className="form">
        <h1 className="form__title">Login</h1>
        <form onSubmit={this.onSubmit} className="uk-form-stacked">
          {this.props.authenticationError && (
            <p className="form__error">{this.props.authenticationError}</p>
          )}
          <label htmlFor="email" className="form__label">
            Email:{" "}
          </label>
          {this.state.errors.email && (
            <p className="form__error">{this.state.errors.email}</p>
          )}
          <input
            type="text"
            name="email"
            value={this.state.email}
            onChange={this.onEmailChange}
            className={`uk-input form__input ${
              this.state.errors.email ? "uk-form-danger" : "uk-form-blank"
            }`}
          />
          <br />
          <label htmlFor="password" className="form__label">
            Password:{" "}
          </label>
          {this.state.errors.password && (
            <p className="form__error">{this.state.errors.password}</p>
          )}
          <input
            type="password"
            name="password"
            value={this.state.password}
            onChange={this.onPasswordChange}
            className={`uk-input form__input ${
              this.state.errors.password ? "uk-form-danger" : "uk-form-blank"
            }`}
          />
          <br />
          <button
            type="submit"
            className="uk-button uk-button-default button button--form"
            disabled={this.props.actionStatus === "Action in progress"}
          >
            {this.props.actionStatus === "Action in progress" ? (
              <Fragment>
                <span>Logging in </span>{" "}
                <FontAwesomeIcon icon={faSpinner} spin />
              </Fragment>
            ) : (
              "Login"
            )}
          </button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  actionStatus: state.status.status,
  authenticationError: state.error.message
});

const mapDispatchToProps = dispatch => ({
  login: userCredentials =>
    dispatch(startAuthentication(userCredentials, "login "))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LoginPage);
