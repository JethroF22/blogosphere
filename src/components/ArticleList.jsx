import React from "react";
import { Link } from "react-router-dom";

import ArticleListItem from "./ArticleListItem";

const ArticleList = props => (
  <div className="article-list">
    {props.posts.map(post => (
      <Link
        to={`/blog/view/${post.slug}`}
        key={Math.floor(Math.random() * 99999)}
      >
        <ArticleListItem post={post} />
      </Link>
    ))}
  </div>
);

export default ArticleList;
