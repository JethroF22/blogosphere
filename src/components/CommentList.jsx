import React from "react";
import CommentListItem from "./CommentListItem";

const CommentList = props => (
  <div className="comments">
    {props.comments.map(comment => <CommentListItem comment={comment} />)}
  </div>
);

export default CommentList;
