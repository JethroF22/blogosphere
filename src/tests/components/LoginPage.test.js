import React from "react";
import { shallow } from "enzyme";
import _ from "lodash";

import { LoginPage } from "../../components/LoginPage";
import users from "../seed/users";

let wrapper, login, token, userCredentials;

beforeEach(() => {
  userCredentials = _.pick(users[0], ["email", "password"]);
  token = "token";
  login = jest.fn().mockResolvedValue(token);
  wrapper = shallow(<LoginPage login={login} />);
  wrapper.setState({ ...userCredentials });
});

test("Should render component correctly", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Should handle login", () => {
  wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  expect(login).toHaveBeenLastCalledWith(userCredentials);
});

test("Should set email on input change", () => {
  const value = "test@gmail.com";
  wrapper
    .find("input")
    .at(0)
    .simulate("change", {
      target: {
        value
      }
    });
  expect(wrapper.state("email")).toBe(value);
});

test("Should set password on input change", () => {
  const value = "password";
  wrapper
    .find("input")
    .at(1)
    .simulate("change", {
      target: {
        value
      }
    });
  expect(wrapper.state("password")).toBe(value);
});
