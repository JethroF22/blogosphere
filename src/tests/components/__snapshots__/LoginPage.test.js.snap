// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render component correctly 1`] = `
<div>
  <h1>
    Login
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="email"
    >
      Email: 
    </label>
    <input
      name="email"
      onChange={[Function]}
      type="text"
      value="userone@gmail.com"
    />
    <br />
    <label
      htmlFor="password"
    >
      Password: 
    </label>
    <input
      name="password"
      onChange={[Function]}
      type="password"
      value="userone1"
    />
    <br />
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
